src/BlockDiagonals.jl

```diff
@@ -0,0 +1,241 @@
+module BlockDiagonals
+
+using BlockArrays
+using BlockArrays: AbstractBlockSizes, BlockSizes
+using FillArrays
+using LinearAlgebra
+
+export BlockDiagonal, blocks
+
+"""
+    BlockDiagonal{T, V<:AbstractMatrix{T}} <: AbstractBlockMatrix{T}
+
+A BlockMatrix with square blocks on the diagonal, and zeros off the diagonal.
+"""
+struct BlockDiagonal{T, S<:AbstractBlockSizes} <: AbstractBlockMatrix{T}
+    blocks::Vector{<:AbstractMatrix{T}}
+    block_sizes::S
+    function BlockDiagonal(blocks::Vector{<:AbstractMatrix{T}}, sizes::S) where {T, S<:AbstractBlockSizes}
+        all(is_square, blocks) || throw(ArgumentError("All blocks must be square."))
+        # mapreduce would give Array of Tuples; want Array of Arrays
+        return new{T, S}(blocks, sizes)
+    end
+end
+
+function BlockDiagonal(blocks::AbstractVector{<:AbstractMatrix})
+    # mapreduce would give Array of Tuples; want Array of Arrays
+    bs = BlockSizes([[first(size(b)), last(size(b))] for b in blocks]...)
+    return BlockDiagonal(blocks, bs)
+end
+
+is_square(A::AbstractMatrix) = size(A, 1) === size(A, 2)
+blocks(b::BlockDiagonal) = b.blocks
+
+# AbstractBlockMatrix interface
+BlockArrays.blocksizes(b::BlockDiagonal) = b.block_sizes
+function BlockArrays.blocksize(b::BlockDiagonal, i::Int, j::Int)
+    return first(size(blocks(b)[i])), last(size(blocks(b)[j]))
+end
+BlockArrays.nblocks(b::BlockDiagonal) = length(blocks(b)), length(blocks(b))
+BlockArrays.nblocks(b::BlockDiagonal, dims::Integer) = length(blocks(b))
+function BlockArrays.getblock(b::BlockDiagonal{T}, i::Int, j::Int) where T
+    return i == j ? blocks(b)[i] : Zeros{T}(blocksize(b, i, j))
+end
+function BlockArrays.setblock!(b::BlockDiagonal{T, V}, v::V, p::Int, q::Int) where {T, V}
+    if p != q
+        throw(brgumentError("Cannot set off-diagonal block ($p, $q) to a nonzero value."))
+    end
+    if blocksize(b, p, q) != size(v)
+        throw(DimensionMismatch(string(
+            "Cannot set block of size $(blocksize(b, p, q)) to block of size $(size(v))"
+        )))
+    end
+    blocks(b)[p] = v
+end
+
+
+Base.Matrix(b::BlockDiagonal) = cat(blocks(b)...; dims=(1, 2))
+Base.size(b::BlockDiagonal) = sum(x -> size(x, 1), blocks(b)), sum(x -> size(x, 2), blocks(b))
+Base.similar(B::BlockDiagonal) = BlockDiagonal(similar.(blocks(B)))
+
+function Base.getindex(b::BlockDiagonal{T}, i::Int, j::Int) where T
+    cols = [size(bb, 2) for bb in blocks(b)]
```

Comment:
```suggestion
    cols = size.(blocks(b), 2)
```

Reply:
THough since it is never mutated maybe should be going for Tuples.

Or LazyArrays